File: sdks/typescript/tests/sync.test.ts
import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
import { LocalStore } from "../src/store.js";
import { SyncManager } from "../src/sync.js";
import type { Gene } from "../src/gene.js";
import type { ExperienceClient } from "../src/experience.js";

function sampleGene(id = "gene-1"): Gene {
  const now = new Date();
  return {
    geneId: id,
    name: `test-${id}`,
    taskClass: "bugfix",
    confidence: 0.85,
    strategy: { approach: "retry" },
    signals: {},
    validation: {},
    qualityScore: 0.9,
    useCount: 5,
    successCount: 4,
    contributorId: "node-1",
    source: "local",
    createdAt: now,
    updatedAt: now,
  };
}

function mockExperience(overrides?: Partial<ExperienceClient>) {
  return {
    share: vi.fn().mockResolvedValue({ gene_id: "gene-1", status: "published" }),
    fetch_: vi.fn().mockResolvedValue({ assets: [], next_cursor: null, sync_audit: { total_available: 0, returned: 0 } }),
    registerPublicKey: vi.fn().mockResolvedValue(undefined),
    ...overrides,
  } as unknown as ExperienceClient;
}

describe("SyncManager", () => {
  let store: LocalStore;

  beforeEach(() => {
    store = new LocalStore(":memory:");
  });
  afterEach(() => {
    store.close();
  });

  it("pushToHub pushes unsynced genes", async () => {
    store.save(sampleGene());
    const exp = mockExperience();
    const mgr = new SyncManager({ store, experience: exp });

    const result = await mgr.pushToHub();
    expect(result.pushed).toBe(1);
    expect(result.failed).toBe(0);
    expect(exp.share).toHaveBeenCalledOnce();

    const got = store.get("gene-1")!;
    expect(got.syncedAt).toBeDefined();

    const logs = store.getSyncLog(10);
    expect(logs).toHaveLength(1);
    expect(logs[0].status).toBe("success");
  });

  it("pushToHub with specific gene ids", async () => {
    store.save(sampleGene("gene-1"));
    store.save(sampleGene("gene-2"));
    const exp = mockExperience();
    const mgr = new SyncManager({ store, experience: exp });

    const result = await mgr.pushToHub({ geneIds: ["gene-1"] });
    expect(result.pushed).toBe(1);
    expect(exp.share).toHaveBeenCalledOnce();
  });

  it("pushToHub handles failure", async () => {
    store.save(sampleGene());
    const exp = mockExperience({
      share: vi.fn().mockRejectedValue(new Error("network error")),
    });
    const mgr = new SyncManager({ store, experience: exp });

    const result = await mgr.pushToHub();
    expect(result.failed).toBe(1);
    expect(result.pushed).toBe(0);
    expect(result.errors[0].message).toContain("network error");
  });

  it("pullFromHub imports new genes", async () => {
    const exp = mockExperience({
      fetch_: vi.fn().mockResolvedValue({
        assets: [
          { type: "gene", id: "remote-1", confidence: 0.9, quality_score: 0.85, use_count: 10, success_count: 8 },
        ],
        next_cursor: null,
        sync_audit: { total_available: 1, returned: 1 },
      }),
    });
    const mgr = new SyncManager({ store, experience: exp });

    const imported = await mgr.pullFromHub({ limit: 10 });
    expect(imported).toBe(1);

    const got = store.get("remote-1")!;
    expect(got.source).toBe("hub");
    expect(got.confidence).toBe(0.9);
  });

  it("pullFromHub skips on conflict", async () => {
    const now = new Date();
    store.save({
      geneId: "remote-1", name: "existing", taskClass: "feature",
      confidence: 0.5, qualityScore: 0, useCount: 0, successCount: 0,
      contributorId: "", source: "local", createdAt: now, updatedAt: now,
    });

    const exp = mockExperience({
      fetch_: vi.fn().mockResolvedValue({
        assets: [{ type: "gene", id: "remote-1", confidence: 0.9 }],
        next_cursor: null,
        sync_audit: { total_available: 1, returned: 1 },
      }),
    });
    const mgr = new SyncManager({ store, experience: exp });

    const imported = await mgr.pullFromHub();
    expect(imported).toBe(0);

    const logs = store.getSyncLog(10);
    expect(logs.some((l) => l.status === "conflict")).toBe(true);
  });

  it("pullFromHub merges stats when same task class", async () => {
    const now = new Date();
    store.save({
      geneId: "gene-1", name: "existing", taskClass: "unknown",
      confidence: 0.5, qualityScore: 0.3, useCount: 3, successCount: 2,
      contributorId: "", source: "local", createdAt: now, updatedAt: now,
    });

    const exp = mockExperience({
      fetch_: vi.fn().mockResolvedValue({
        assets: [{ type: "gene", id: "gene-1", confidence: 0.9, quality_score: 0.8, use_count: 10, success_count: 8 }],
        next_cursor: null,
        sync_audit: { total_available: 1, returned: 1 },
      }),
    });
    const mgr = new SyncManager({ store, experience: exp });

    const imported = await mgr.pullFromHub();
    expect(imported).toBe(1);

    const got = store.get("gene-1")!;
    expect(got.confidence).toBe(0.9);
    expect(got.useCount).toBe(10);
    expect(got.successCount).toBe(8);
  });

  it("throws when no experience client", async () => {
    const mgr = new SyncManager({ store });
    await expect(mgr.pushToHub()).rejects.toThrow("experience client not configured");
    await expect(mgr.pullFromHub()).rejects.toThrow("experience client not configured");
  });
});

File: sdks/go/sync/sync_test.go
package sync

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/Colin4k1024/Oris/sdks/go/experience"
	"github.com/Colin4k1024/Oris/sdks/go/store"
)

func newTestStore(t *testing.T) *store.LocalStore {
	t.Helper()
	s, err := store.Open(":memory:")
	if err != nil {
		t.Fatalf("open store: %v", err)
	}
	t.Cleanup(func() { s.Close() })
	return s
}

func TestPushToHub(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodPost && r.URL.Path == "/experience" {
			json.NewEncoder(w).Encode(experience.ShareResponse{
				GeneID:      "gene-1",
				Status:      "published",
				PublishedAt: time.Now(),
			})
			return
		}
		w.WriteHeader(http.StatusNotFound)
	}))
	defer server.Close()

	s := newTestStore(t)
	ctx := context.Background()

	now := time.Now().UTC()
	g := store.Gene{
		GeneID:    "gene-1",
		Name:      "test-gene",
		TaskClass: "bugfix",
		Confidence: 0.85,
		Source:    "local",
		CreatedAt: now,
		UpdatedAt: now,
	}
	s.Save(ctx, g)

	expClient := experience.NewClient(experience.Config{
		BaseURL:    server.URL,
		APIKey:     "test-key",
		SenderID:   "test-node",
		HTTPClient: server.Client(),
	})

	mgr := New(Config{Store: s, Experience: expClient})
	result, err := mgr.PushToHub(ctx, store.PushOpts{})
	if err != nil {
		t.Fatalf("push: %v", err)
	}
	if result.Pushed != 1 {
		t.Errorf("pushed = %d, want 1", result.Pushed)
	}
	if result.Failed != 0 {
		t.Errorf("failed = %d, want 0", result.Failed)
	}

	got, _ := s.Get(ctx, "gene-1")
	if got.SyncedAt == nil {
		t.Error("gene should be marked as synced")
	}

	logs, _ := s.GetSyncLog(ctx, 10)
	if len(logs) != 1 {
		t.Fatalf("sync log len = %d, want 1", len(logs))
	}
	if logs[0].Status != "success" {
		t.Errorf("log status = %q, want success", logs[0].Status)
	}
}

func TestPushToHubFailure(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusInternalServerError)
		w.Write([]byte("internal error"))
	}))
	defer server.Close()

	s := newTestStore(t)
	ctx := context.Background()

	now := time.Now().UTC()
	s.Save(ctx, store.Gene{
		GeneID:    "gene-1",
		Name:      "test",
		TaskClass: "bugfix",
		Confidence: 0.5,
		Source:    "local",
		CreatedAt: now,
		UpdatedAt: now,
	})

	expClient := experience.NewClient(experience.Config{
		BaseURL:    server.URL,
		APIKey:     "test-key",
		SenderID:   "test-node",
		HTTPClient: server.Client(),
	})

	mgr := New(Config{Store: s, Experience: expClient})
	result, err := mgr.PushToHub(ctx, store.PushOpts{})
	if err != nil {
		t.Fatalf("push: %v", err)
	}
	if result.Failed != 1 {
		t.Errorf("failed = %d, want 1", result.Failed)
	}
	if result.Pushed != 0 {
		t.Errorf("pushed = %d, want 0", result.Pushed)
	}
	if len(result.Errors) != 1 {
		t.Fatalf("errors len = %d, want 1", len(result.Errors))
	}
}

func TestPullFromHub(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodGet && r.URL.Path == "/experience" {
			resp := experience.FetchResponse{
				Assets: []experience.NetworkAsset{
					{
						Type:         "gene",
						ID:           "remote-gene-1",
						Confidence:   0.9,
						QualityScore: 0.85,
						UseCount:     10,
						SuccessCount: 8,
						CreatedAt:    time.Now().UTC(),
					},
				},
				SyncAudit: experience.SyncAudit{TotalAvailable: 1, Returned: 1},
			}
			json.NewEncoder(w).Encode(resp)
			return
		}
		w.WriteHeader(http.StatusNotFound)
	}))
	defer server.Close()

	s := newTestStore(t)
	ctx := context.Background()

	expClient := experience.NewClient(experience.Config{
		BaseURL:    server.URL,
		APIKey:     "test-key",
		SenderID:   "test-node",
		HTTPClient: server.Client(),
	})

	mgr := New(Config{Store: s, Experience: expClient})
	imported, err := mgr.PullFromHub(ctx, store.PullOpts{Limit: 10})
	if err != nil {
		t.Fatalf("pull: %v", err)
	}
	if imported != 1 {
		t.Errorf("imported = %d, want 1", imported)
	}

	got, _ := s.Get(ctx, "remote-gene-1")
	if got == nil {
		t.Fatal("expected imported gene")
	}
	if got.Source != "hub" {
		t.Errorf("source = %q, want hub", got.Source)
	}
	if got.Confidence != 0.9 {
		t.Errorf("confidence = %f, want 0.9", got.Confidence)
	}
}

func TestPullConflictSkipsWhenDifferent(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		resp := experience.FetchResponse{
			Assets: []experience.NetworkAsset{
				{
					Type:       "gene",
					ID:         "gene-1",
					Confidence: 0.9,
				},
			},
		}
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	s := newTestStore(t)
	ctx := context.Background()

	now := time.Now().UTC()
	s.Save(ctx, store.Gene{
		GeneID:    "gene-1",
		Name:      "existing",
		TaskClass: "feature",
		Confidence: 0.5,
		Source:    "local",
		CreatedAt: now,
		UpdatedAt: now,
	})

	expClient := experience.NewClient(experience.Config{
		BaseURL:    server.URL,
		SenderID:   "test",
		HTTPClient: server.Client(),
	})

	mgr := New(Config{Store: s, Experience: expClient})
	imported, _ := mgr.PullFromHub(ctx, store.PullOpts{})
	if imported != 0 {
		t.Errorf("imported = %d, want 0 (conflict should skip)", imported)
	}

	logs, _ := s.GetSyncLog(ctx, 10)
	found := false
	for _, l := range logs {
		if l.Status == "conflict" {
			found = true
		}
	}
	if !found {
		t.Error("expected conflict log entry")
	}
}

func TestNoExperienceClient(t *testing.T) {
	s := newTestStore(t)
	mgr := New(Config{Store: s})

	_, err := mgr.PushToHub(context.Background(), store.PushOpts{})
	if err == nil {
		t.Error("expected error for push without experience client")
	}

	_, err = mgr.PullFromHub(context.Background(), store.PullOpts{})
	if err == nil {
		t.Error("expected error for pull without experience client")
	}
}
